src/State/ForValues.lua

```diff
@@ -0,0 +1,325 @@
+end
+
+--[[
+	Called when the original table is changed.
+
+	This will firstly find any values meeting any of the following criteria:
+
+	- they were not previously present
+	- a dependency used during generation of this value has changed
+
+	It will recalculate those values, storing information about any dependencies
+    used in the processor callback during value generation, and save the new value
+    to the output array with the same key. If it is overwriting an older value,
+	that older value will be passed to the destructor for cleanup.
+
+	Finally, this function will find values that are no longer present, and remove
+	their values from the output table and pass them to the destructor. You can re-use
+    the same value multiple times and this will function will update them as little as
+    possible; reusing the same values where possible.
+]]
+function class:update(): boolean
+	local inputIsState = self._inputIsState
+	local inputValues = self._inputTable
+	local newValueCache = self._oldValueCache -- _oldValueCache and _valueCache get swapped and cleared below
+	local oldValueCache = self._valueCache
+	local dependenciesCaptured = {} -- we use this to cache dependency checks/captures
+	local outputValues = {}
+	local meta = self._meta
+
+	if inputIsState then
+		inputValues = inputValues:get(false)
+	end
+
+	local didChange = false
+
+	-- clean out main dependency set
+	for dependency in pairs(self.dependencySet) do
+		dependency.dependentSet[self] = nil
+	end
+	self._oldDependencySet, self.dependencySet = self.dependencySet, self._oldDependencySet
+	table.clear(self.dependencySet)
+
+	-- swap caches and clear out the old value cache
+	self._oldValueCache, self._valueCache = oldValueCache, newValueCache
+	table.clear(newValueCache)
+
+	-- if the input table is a state object, add as dependency
+	if inputIsState then
+		self._inputTable.dependentSet[self] = true
+		self.dependencySet[self._inputTable] = true
+	end
+
+	-- STEP 1: find values that changed or were not previously present
+	for _key, inValue in pairs(inputValues) do
+		local cachedValue = oldValueCache[inValue]
+		local newCachedValue = newValueCache[inValue]
+		local shouldRecalculate = cachedValue == nil
+
+		-- get value data
+		local valueData = self._valueData[inValue]
+		if valueData == nil then
+			valueData = {
+				-- we don't need strong references here - the main set does that
+				-- for us, so let's not introduce unnecessary leak opportunities
+				dependencySet = setmetatable({}, WEAK_KEYS_METATABLE),
+				oldDependencySet = setmetatable({}, WEAK_KEYS_METATABLE),
+				dependencyValues = setmetatable({}, WEAK_KEYS_METATABLE),
+			}
+			self._valueData[inValue] = valueData
+		end
+
+		-- if we already cached a new constant (non-table) value on this run-through,
+		-- then we can just re-use that as the cached value
+		if newCachedValue ~= nil and type(newCachedValue) ~= "table" then
+			cachedValue = newCachedValue
+			shouldRecalculate = false
```

Comment:
This is unnecessary, because shouldRecalculate is only initialised as true if there is no cached value. Leaving this here may introduce logic bugs later if shouldRecalculate gets assigned earlier in later commits.

Reply:
This enables us to use the newly cached constant without running additional processor functions. 

Consider this: We do not have a cached value from the previous run, but we cached a new constant value on this run. In this case ``shouldRecalculate`` will be true.  If we don't explicitly set ``shouldRecalculate`` to false here,  the processor will re-run (because we will go into the if block on line 196), unnecessarily recalculating the constant value we already calculated. This creates a mismatch in the number of processor and destructor calls.

Reply:
I mistook newCachedValue for the original cachedValue. So nevermind.